### PR TITLE
Overhaul gather()

### LINKED ISSUE #4698: Client/Scheduler gather not robust to busy worker
When I run a lot of tasks on the CNES HPC with a big Dask cluster (512 threads/128 workers), I sometimes have communication errors between the scheduler and the workers. The error is thrown by the module "distributed/utils_comm.py," because the code tries to read the key 'data' which does not exist in the dictionary.

```python
    /home/ad/briolf/anaconda3/lib/python3.8/site-packages/distributed/utils_comm.py in retry_operation
          383         dask.config.get("distributed.comm.retry.delay.max"), default="s"
          384     )
    ----> 385     return await retry(
          386         partial(coro, *args, **kwargs),
          387         count=retry_count,
    
    /home/ad/briolf/anaconda3/lib/python3.8/site-packages/distributed/utils_comm.py in retry
          368                 delay *= 1 + random.random() * jitter_fraction
          369             await asyncio.sleep(delay)
    ----> 370     return await coro()
          371 
          372 
    
    /home/ad/briolf/anaconda3/lib/python3.8/site-packages/distributed/core.py in send_recv_from_rpc
          860             name, comm.name = comm.name, "ConnectionPool." + key
          861             try:
    ----> 862                 result = await send_recv(comm=comm, op=key, **kwargs)
          863             finally:
          864                 self.pool.reuse(self.addr, comm)
    
    /home/ad/briolf/anaconda3/lib/python3.8/site-packages/distributed/core.py in send_recv
          659         if comm.deserialize:
          660             typ, exc, tb = clean_exception(**response)
    ----> 661             raise exc.with_traceback(tb)
          662         else:
          663             raise Exception(response["text"])
    
    /home/ad/briolf/anaconda3/lib/python3.8/site-packages/distributed/core.py in handle_comm
          499                             result = asyncio.ensure_future(result)
          500                             self._ongoing_coroutines.add(result)
    ----> 501                             result = await result
          502                     except (CommClosedError, CancelledError) as e:
          503                         if self.status == Status.running:
    
    /home/ad/briolf/anaconda3/lib/python3.8/site-packages/distributed/scheduler.py in gather
          5052                 who_has[key] = []
          5053 
    ----> 5054         data, missing_keys, missing_workers = await gather_from_workers(
          5055             who_has, rpc=self.rpc, close=False, serializers=serializers
          5056         )
    
    /home/ad/briolf/anaconda3/lib/python3.8/site-packages/distributed/utils_comm.py in gather_from_workers
          86                     missing_workers.add(worker)
          87                 else:
    ----> 88                     response.update(r["data"])
          89         finally:
          90             for r in rpcs.values():
    
    KeyError: 'data'
```
I modified the module code, to display the content of this dictionary when this error is thrown. I saw that in this context, the dictionary just contains the following data: ```dict(status='busy')```.

I cannot put reproducible code because the software used is not public and I did not discover simple lines of code to reproduce this problem. I can eventually do other tests to give you further information.

**Environment**:

- Dask version: 2021.4.0
- Python version: 3.8.8
- Operating System: CentOS Linux release 7.6.1810
- Install method (conda, pip, source): conda, channel conda-forge

### LINKED ISSUE #7982: Race condition between gather() and AMM
- XREF #4698
- XREF #7993 
- XREF #7995
- XREF #7996

The `who_has` mapping used by `gather` can fall out of sync.
This can happen, for example:

1. when the client is gathering intermediate keys, that are dependencies to other keys, and because of this they are replicated on additional workers while `Scheduler.gather` is running. Afterwards, AMM ReduceReplicas deletes one or more of the excess replicas, which may be the ones in the `who_has` mapping.
2. when a dynamic cluster is shrinking down, which in turn is causing AMM RetireWorkers to create replicas somewhere else and delete the old ones.

When this happens, if a remote worker is missing *any* key, `gather` completely removes the worker from its own internal copy of who_has. It does not refresh who_has from the scheduler.

Once the local copy of `who_has` owned by `gather` runs out of workers for a key, the key is marked as missing, and `gather` fails .

### LINKED ISSUE #7993: gather(direct=True) may fall back to direct=False
The current implementation for

-  `Client(...).gather(..., direct=True)` 
- `Client(direct_to_workers=True).gather(...)`
- `get_client().gather(...)` (worker clients)
 
is that, if *anything* bad happens on the direct connection to the workers, the keys that failed to be fetched directly are fetched through the scheduler `(direct=False)` instead.

I believe that the intent of this design is to build resilience when the user specifies direct=True while in fact they have a firewall preventing comms.

First, I think even the original intent is a very bad idea to begin with - most firewalls will just silently drop the packets, thus causing the user to wait 30(?) seconds for timeout and not figuring out why it's taking so long.

Second, there are many cases where a key will fail to be gathered on the first try - notably, the who_has dict can fall out of sync with the scheduler. This may cause a workflow that gathers huge amounts of data from workers to client to run fine *most times* and then, once in a while, kill off the scheduler due to memory pressure.

I suggest we drop this feature completely and instead build resilience to changes in who_has (as described in #7982). If workers are unreachable with direct=True, gather should just fail. Note that the default is direct=False.

CC @fjetter @hendrikmakait

### LINKED ISSUE #7995: gather() should not remove unresponsive workers
`Scheduler.gather` removes workers it fails to connect to.
However, if it fails to connect to a worker, but then successfully retrieves the same key from another worker, this will not happen.

This is a bad idea to begin with, and the last remnant of an old design where various failed RPC calls would shut down workers.
One key failure point is that temporary GIL congestion or a minor glitch in the network can cause new RPC calls to randomly fail.
The current philosophy is that only a collapse of the Batched Comms - which, once established, is kept alive by the kernel and is immune to the GIL - should cause worker and scheduler to get disconnected.

### LINKED ISSUE #7996: gather() prints confusing error messages
`Scheduler.gather` prints confusing error messages:
```
distributed.scheduler - ERROR - Shut down workers that don't have promised key: [], ('to_crs-24a4e1b9527e04817a29817fb49c2ca7', 148)
NoneType: None
distributed.scheduler - ERROR - Shut down workers that don't have promised key: ['tls://10.1.89.83:33685'], ('to_crs-24a4e1b9527e04817a29817fb49c2ca7', 79)
NoneType: None
```
This is wrong. What the scheduler is actually doing is shutting down workers that have been found to be completely unresponsive. They potentially may not appear anywhere in the log.

Additionally, the workers reported to be unresponsive may have actually have been already removed earlier, maybe gracefully - e.g. because the cluster was shrinking down.

### PR DESCRIPTION
**Pull request by crusaderky**
- Closes #4698
- Closes #7982
- Closes #7996
- Supersedes #5546
- DOES NOT address #7995
- DOES NOT address #7993
- Soft-blocked by #8006

### CODE REVIEW

File: distributed/client.py [hunk 1]
```diff
@@ -2311,22 +2311,26 @@ async def _gather_remote(self, direct, local_worker):
...
+            data, missing_keys = await gather_from_workers(
+                keys=keys, who_has=who_has, rpc=self.rpc
+            )
+            response: dict[str, Any] = {"status": "OK", "data": data}
+            if missing_keys:
+                response = await retry_operation(
+                    self.scheduler.gather, keys=missing_keys
+                )
+                if response["status"] == "OK":
+                    response["data"].update(data)
```
Comment (crusaderky) [inline distributed/client.py:2331]: - #7993 
This PR does not change the previous behaviour.

File: distributed/utils_comm.py [hunk 2]
```diff
@@ -77,36 +116,46 @@ async def gather_from_workers(
...
+        for address, task in tasks.items():
             try:
-                r = await c
-            except OSError:
-                missing_workers.add(worker)
-            except ValueError as e:
-                logger.info(
-                    "Got an unexpected error while collecting from workers: %s", e
+                r = await task
+            except (OSError, asyncio.CancelledError, asyncio.TimeoutError):
```
Comment (crusaderky) [inline distributed/utils_comm.py:129]: TODO test (blocked by #8013)

File: distributed/tests/test_utils_comm.py [hunk 3]
```diff
@@ -63,10 +61,57 @@ async def test_gather_from_workers_permissive_flaky(c, s, a, b):
...
+async def test_gather_from_workers_cancelled_error(c, s):
+    """Something somewhere in the networking stack raises CancelledError while
+    gather_from_workers is running
+
+    See Also
+    --------
+    test_worker.py::test_gather_dep_cancelled_error
+    test_worker.py::test_get_data_cancelled_error
+    https://github.com/dask/distributed/issues/8006
+    """
```
Comment (crusaderky) [inline distributed/tests/test_utils_comm.py:80]: XREF #8013

**Comment by crusaderky** [comment]
This is ready for review however I'd rather have #8013 in before so that we don't muddle the waters.

File: distributed/utils_comm.py [hunk 4]
```diff
@@ -77,36 +116,46 @@ async def gather_from_workers(
...
+                # For example, deserialization error
+                logger.exception(
+                    "Unexpected error while collecting tasks %s from %s",
+                    d[address],
+                    address,
                 )
-                missing_workers.add(worker)
+                for key in d[address]:
+                    missing_keys.add(key)
+                    del to_gather[key]
```
Comment (hendrikmakait) [inline distributed/utils_comm.py:143]: Potentially only one of these keys caused the exception. Would it make sense to attempt gathering them individually to keep the # missing keys to a minimum? 

We may also consider returning these as erring keys, not missing. However, this is out of scope.

File: distributed/worker.py [hunk 5]
```diff
@@ -1307,23 +1306,31 @@ def keys(self) -> list[str]:
...
+        async def refresh_who_has(keys: list[str]) -> Mapping[str, Collection[str]]:
+            nonlocal first
+            if first:
+                first = False
+                return who_has
+            return await retry_operation(self.scheduler.who_has, keys=keys)
+
+        result, missing_keys = await gather_from_workers(
+            keys=who_has,
+            who_has=refresh_who_has,
```
Comment (hendrikmakait) [inline distributed/worker.py:1326]: As noted above, the interchangeable use of `who_has` is confusing. Renaming the parameter of `gather_from_workers` to `refresh_who_has` might help.

File: distributed/worker.py [hunk 6]
```diff
@@ -1707,17 +1714,11 @@ async def batched_send_connect():
...
-        if max_connections is None:
-            max_connections = self.transfer_outgoing_count_limit
-
-        if keys is None:
-            keys = set()
-
+        max_connections = self.transfer_outgoing_count_limit
         # Allow same-host connections more liberally
         if (
             max_connections
```
Comment (hendrikmakait) [inline distributed/worker.py:1724]: nit: IIUC, max_connections is now always an integer, so this can be dropped. The same goes for line 1738/1739 `max_connections is not False`.

File: distributed/utils_comm.py [hunk 7]
```diff
@@ -22,52 +30,83 @@
 
 
 async def gather_from_workers(
-    who_has: Mapping[str, Collection[str]],
+    keys: Iterable[str],
+    who_has: Callable[
+        [list[str]],
+        Mapping[str, Collection[str]] | Awaitable[Mapping[str, Collection[str]]],
+    ],
```
Comment (fjetter) [inline distributed/utils_comm.py:37]: I'm not convinced this callback structure is the right approach here. I find this makes the entire mechanism even harder to reason about. Besides, this bypasses (for better or worse) missing/update_who_has mechanics on the worker.
I believe this kind of logic should be handled a layer further up the stack instead of throwing this all into this function.

File: distributed/worker.py [hunk 8]
```diff
@@ -1307,23 +1306,31 @@ def keys(self) -> list[str]:
...
             for k, v in who_has.items()
             if k not in self.data
         }
-        result, missing_keys, missing_workers = await gather_from_workers(
-            who_has=who_has, rpc=self.rpc, who=self.address
+
+        async def refresh_who_has(keys: list[str]) -> Mapping[str, Collection[str]]:
+            nonlocal first
+            if first:
+                first = False
```
Comment (fjetter) [inline distributed/worker.py:1320]: why this? Isn't the point of this callback architecture to be able to update this on every iteration?

File: distributed/tests/test_utils_comm.py (hunk 3 above)
Comment (fjetter) [reply distributed/tests/test_utils_comm.py:82]: How is this test related to the PR?
I'm rather conserned about this test. I don't want us to artificially cancel the tasks. The "network stack" should not be raising a CancelledError at all without a reason. Catching those can have unwanted and hard to debug consequences. This test

File: distributed/tests/test_scheduler.py [hunk 9]
```diff
@@ -2855,56 +2855,31 @@ async def test_gather_no_workers(c, s, a, b):
...
+    with captured_logger("distributed.scheduler") as sched_logger:
+        with captured_logger("distributed.client") as client_logger:
+            assert await c.gather(x, direct=False) == 2
 
-            assert s.tasks[fin.key].who_has == {s.workers[b.address]}
-            assert a.state.executed_count == 2
-            assert b.state.executed_count >= 1
-            # ^ leave room for a future switch from `remove_worker` to `retire_workers`
+    assert sched_logger.getvalue() == "Couldn't gather keys: {'x': 'memory'}\n" * 3
+    assert "Couldn't gather 1 keys, rescheduling" in client_logger.getvalue()
```
Comment (fjetter) [inline distributed/tests/test_scheduler.py:2878]: Putting aside for a moment what changes you are proposing here. Just reading this test I believe this behavior is false.

This test is fetching data via the scheduler but is running into connection failures. However, the `Worker` is still alive (otherwise the `BatchedSend` would've been broken and the `Worker` removed). Despite of knowing that the `Worker` is alive and it merely struggles to connect, it is *rescheduling* the key? This feels like an unwarranted escalation.

**Review by fjetter — COMMENTED** [review]
There is a lot happening in this PR. Can we break this up a bit?

File: distributed/utils_comm.py [hunk 10]
```diff
@@ -22,52 +30,83 @@
...
+            addresses -= busy_workers
+            if addresses:
+                d[random.choice(list(addresses))].append(key)
+
+        if not d:
+            if busy_workers:
+                await asyncio.sleep(0.15)
+                busy_workers.clear()
+
+            new_who_has = who_has(list(to_gather))
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/utils_comm.py:89]: I'll rename it to get_who_has to match Scheduler.get_who_has

File: distributed/utils_comm.py (hunk 7 above)
Comment (crusaderky) [reply distributed/utils_comm.py:37]: The method on the worker exclusively serves replicate, which we need to pen in to replace with AMM to begin with. The original method bypasses the whole worker missing/update_who_has system and I chose not to change that logic since it would be very complicated and ultimately throwaway.

File: distributed/worker.py (hunk 8 above)
Comment (crusaderky) [reply distributed/worker.py:1320]: This avoids introducing an unnecessary additional round-trip from worker to scheduler every time the user calls replicate().
If you think that code cleanliness trumps performance in this very niche use case, I'm happy to simplify the logic.

File: distributed/client.py (hunk 1 above)
Comment (crusaderky) [reply distributed/client.py:2331]: I cannot see any aesthetical refactoring in this PR that could be moved to a separate PR while preserving the functionality.

File: distributed/utils_comm.py [hunk 11]
```diff
@@ -77,36 +116,46 @@ async def gather_from_workers(
...
-                missing_workers.add(worker)
-            except ValueError as e:
-                logger.info(
-                    "Got an unexpected error while collecting from workers: %s", e
+                r = await task
+            except (OSError, asyncio.CancelledError, asyncio.TimeoutError):
+                # Note: CancelledError and asyncio.TimeoutError are rare conditions
+                # that can be raised by the network stack.
+                # See https://github.com/dask/distributed/issues/8006
+                missing_workers.add(address)
```
Comment (fjetter) [inline distributed/utils_comm.py:133]: Catching this `CancelledError` is actually slightly worse than https://github.com/dask/distributed/pull/8013
In the other PR we were talking about a well isolated task and ensured that this task could still terminate safely when it caught a CancelledError. In this case, `gather_data_from_workers` is embedded in another coroutine and a CancelledError would actually not cause this stack to unravel but would instead cause a chain of events to handle `missing_workers`.

File: distributed/tests/test_utils_comm.py [hunk 12]
```diff
@@ -64,10 +63,58 @@ async def test_gather_from_workers_permissive_flaky(c, s, a, b):
...
+        assert tasks
+        # There should be only one task but cope with finding more just in case a
+        # previous test didn't properly clean up
+        for task in tasks:
+            task.cancel()
+
+        a.block_get_data.set()
+        b.block_get_data.set()
+        # gather_from_workers retries transparently from the other worker
+        assert await fut == ({"x": 1}, set())
```
Comment (fjetter) [inline distributed/tests/test_utils_comm.py:117]: This test actually shows that a `CancelledError` indeed does not cause the coroutines to abort but rather that something else is happening. This is actually the behavior I am concerned about.

I suggest to move this to a different PR

**Comment by crusaderky** [comment]
As discussed online and offline:
- moved fix of #7995 to a future PR
- moved handling of CancelledError to match #8013 to a future PR
- refactored implementation without the callback
- removed cosmetic refactorings as much as possible

**Comment by crusaderky** [comment]
@hendrikmakait @fjetter 
This is in theory complete and ready for a second round of review.
However, I'm seeing a lot of instability in CI.

These failures I've never seen before; I'll investigate:
- 1 out of 10 runs failed: test_TaskStreamPlugin (distributed.diagnostics.tests.test_task_stream)
- 1 out of 10 runs failed: test_gather_after_failed_worker (distributed.tests.test_failed_workers)
- 1 out of 10 runs failed: test_dont_steal_fast_tasks_compute_time (distributed.tests.test_steal)

These are known offenders but potentially impacted by this PR; I'll investigate:
- 1 out of 10 runs failed: test_gather_then_submit_after_failed_workers (distributed.tests.test_failed_workers)
- 1 out of 10 runs failed: test_tell_workers_when_peers_have_left (distributed.tests.test_scheduler)
- 2 out of 10 runs failed: test_chaos_rechunk (distributed.tests.test_stress)

These are known offenders which shouldn't be impacted:
- 1 out of 8 runs failed: test_closed_input_only_worker_during_transfer (distributed.shuffle.tests.test_shuffle)
- 1 out of 8 runs failed: test_closed_worker_during_transfer (distributed.shuffle.tests.test_shuffle)
- 1 out of 8 runs failed: test_crashed_worker_during_transfer (distributed.shuffle.tests.test_shuffle)
- 2 out of 8 runs failed: test_restarting_during_transfer_raises_killed_worker (distributed.shuffle.tests.test_shuffle)
- 1 out of 10 runs failed: test_file_descriptors_dont_leak[Nanny] (distributed.tests.test_client)

File: distributed/utils_comm.py (hunk 4 above)
Comment (crusaderky) [reply distributed/utils_comm.py]: Changed code to mark keys as erring, meaning that gather() won't try again.
However, note that there is no transition memory->error in the scheduler; this is the same problem as with gather_dep (https://github.com/dask/distributed/issues/6705).

Yes, we *could* try a fetch the keys one by one, but it feels like a substantial complication and ultimately overkill for what, *in theory*, should be a rare-ish problem easily weeded out during development?

File: distributed/client.py [hunk 13]
```diff
@@ -2311,22 +2311,26 @@ async def _gather_remote(self, direct, local_worker):
...
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/client.py]: I would like to get rid of the whole functionality in #7993, so it would be throwaway work.

**Comment by crusaderky** [comment]
I've run all tests above 10 times on each CI environment, on main and on this PR, and counted the number of failures, and I can see no regression; this is ready for final review and merge.

On a side note, this was an extremely time-consuming activity and we should urgently pen in the time to fix CI.


| test | Failures in main | Failures in this PR |
| --- | --- | --- |
| diagnostics/tests/test_task_stream.py::test_TaskStreamPlugin | 1 | 0 |
| shuffle/tests/test_shuffle.py::test_clean_after_close | 1 | 0 |
| shuffle/tests/test_shuffle.py::test_closed_input_only_worker_during_transfer | 0 | 1 |
| shuffle/tests/test_shuffle.py::test_closed_worker_during_transfer | 29 | 35 |
| shuffle/tests/test_shuffle.py::test_crashed_worker_during_transfer | 6 | 1 |
| shuffle/tests/test_shuffle.py::test_restarting_during_transfer_raises_killed_worker | 38 | 33 |
| tests/test_client.py::test_file_descriptors_dont_leak | 9 | 10 |
| tests/test_failed_workers.py::test_gather_after_failed_worker | 4 | 2 |
| tests/test_failed_workers.py::test_gather_then_submit_after_failed_workers | 9 | 1 |
| tests/test_scheduler.py::test_tell_workers_when_peers_have_left | 1 | 1 |
| tests/test_stress.py::test_chaos_rechunk | 0 | 1 |

File: distributed/utils_comm.py [hunk 14]
```diff
@@ -77,36 +86,42 @@ async def gather_from_workers(
...
                 name=f"get-data-from-{address}",
             )
             for address, keys in d.items()
         }
-        response: dict[str, object] = {}
-        for worker, c in coroutines.items():
+        for address, task in tasks.items():
             try:
-                r = await c
+                r = await task
```
Comment (fjetter) [inline distributed/utils_comm.py:98]: I know this is the same as before but the task scheduling pattern here is problematic in the event of cancellation.

If `gather_from_workers` is cancelled after the tasks were created, only the task we're currently awaiting is cancelled, all others will continue running and we'll get a "never awaited foo" warning.

I _guess_ this coroutinefunction is never actually cancelled which is why we never ran into this...

The _correct_ approach to this would be to use `asyncio.gather` (or even better but not backwards copatible, the [TaskGroup of asyncio in python 3.11+](https://docs.python.org/3/library/asyncio-task.html#asyncio.TaskGroup))

I think the changes should be straight forward to something like

```python
results = asyncio.gather(tasks, return_exceptions=True)
for addr, res in results.items():
    if isinstance(res, OSError):
        ...
```

**Review by fjetter — APPROVED** [review]
The above comment is not essential. I think it's good practice to cleanup our asyncio game but considering this hasn't bothered us before, I assume this thing is just never cancelled.

File: distributed/utils_comm.py (hunk 14 above)
Comment (crusaderky) [reply distributed/utils_comm.py:98]: I'll address it in a follow-up
